src/runtime-rs/crates/resource/src/volume/share_fs_volume.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{path::Path, sync::Arc};
+
+use anyhow::{anyhow, Context, Result};
+use log::debug;
+use nix::sys::stat::{stat, SFlag};
+
+use super::Volume;
+use crate::share_fs::{ShareFs, ShareFsVolumeConfig};
+
+// copy file to container's rootfs if filesystem sharing is not supported, otherwise
+// bind mount it in the shared directory.
+// Ignore /dev, directories and all other device files. We handle
+// only regular files in /dev. It does not make sense to pass the host
+// device nodes to the guest.
+// skip the ulog-fs volume
+// skip the volumes whose source had already set to guest share dir.
+pub(crate) struct ShareFsVolume {
+    mounts: Vec<oci::Mount>,
+}
+
+impl ShareFsVolume {
+    pub(crate) async fn new(
+        share_fs: &Option<Arc<dyn ShareFs>>,
+        m: &oci::Mount,
+        cid: &str,
+    ) -> Result<Self> {
+        let file_name = Path::new(&m.source).file_name().unwrap().to_str().unwrap();
+        let f_name = generate_mount_path(cid, file_name);
+
+        let mut volume = Self { mounts: vec![] };
+        let mut need_copy = false;
+        if share_fs.is_none() {
+            match stat(Path::new(&m.source)) {
+                Ok(stat) => {
+                    // Ignore the mount if this is not a regular file (excludes
+                    // directory, socket, device, ...) as it cannot be handled by
+                    // a simple copy. But this should not be treated as an error,
+                    // only as a limitation.
+                    // golang implement:
+                    // ModeType = ModeDir | ModeSymlink | ModeNamedPipe | ModeSocket |
+                    //            ModeDevice | ModeCharDevice | ModeIrregular
+                    let file_type = SFlag::S_IFDIR
+                        | SFlag::S_IFLNK
+                        | SFlag::S_IFIFO
+                        | SFlag::S_IFSOCK
+                        | SFlag::S_IFCHR
+                        | SFlag::S_IFREG;
+                    if !file_type.contains(SFlag::from_bits_truncate(stat.st_mode)) {
+                        debug!(
+                            "Ignoring non-regular file as FS sharing not supported. mount: {:?}",
+                            m
+                        );
+                        return Ok(volume);
+                    }
+                    if SFlag::from_bits_truncate(stat.st_mode) != SFlag::S_IFDIR {
+                        need_copy = true;
+                    }
+                }
+                Err(err) => {
+                    return Err(anyhow!(format!(
+                        "failed to stat file {} {:?}",
+                        &m.source, err
+                    )));
+                }
+            };
+        }
+
+        if need_copy {
```

Comment:
The `need_copy` should only occur when `share_fs.is_none()`, it's not the same dimension, it should only in the `if share_fs.is_none() {}` block.